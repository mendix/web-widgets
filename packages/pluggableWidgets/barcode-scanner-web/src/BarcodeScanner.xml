<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.barcodescanner.BarcodeScanner" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd" pluginWidget="true">
    <name>Barcode Scanner</name>
    <description>The widget lets you scan a barcode</description>
    <studioProCategory>Images, videos &amp; files</studioProCategory>
    <studioCategory>Images, Videos &amp; Files</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/barcode-scanner</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="datasource" type="attribute" required="true">
                    <caption>Scanned result</caption>
                    <description>The String attribute used to store the result of the scanned barcode.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
                <property key="showMask" type="boolean" required="true" defaultValue="true">
                    <caption>Show barcode mask</caption>
                    <description>Apply a mask to camera view, as a specific target area for the barcode.</description>
                </property>
                <property key="useAllFormats" type="boolean" required="true" defaultValue="true">
                    <caption>Use all barcode formats</caption>
                    <description>Scan for all available barcode formats</description>
                </property>
                <property key="barcodeFormats" type="object" isList="true">
                    <caption>Enabled barcode formats</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="Object list group">
                            <property key="barcodeFormat" type="enumeration" defaultValue="AZTEC">
                                <caption>Barcode format</caption>
                                <description>Barcode format which should be recognized by the scanner</description>
                                <enumerationValues>
                                    <enumerationValue key="AZTEC">Aztec</enumerationValue>
                                    <enumerationValue key="CODE_39">Code 39</enumerationValue>
                                    <enumerationValue key="CODE_128">Code 128</enumerationValue>
                                    <enumerationValue key="DATA_MATRIX">Data Matrix</enumerationValue>
                                    <enumerationValue key="EAN_8">EAN-8</enumerationValue>
                                    <enumerationValue key="EAN_13">EAN-13</enumerationValue>
                                    <enumerationValue key="ITF">ITF</enumerationValue>
                                    <enumerationValue key="PDF_417">PDF 417</enumerationValue>
                                    <enumerationValue key="QR_CODE">QR Code</enumerationValue>
                                    <enumerationValue key="RSS_14">RSS-14</enumerationValue>
                                    <enumerationValue key="UPC_A">UPC-A</enumerationValue>
                                    <enumerationValue key="UPC_E">UPC-E</enumerationValue>
                                </enumerationValues>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onDetect" type="action" required="false">
                    <caption>On detect action</caption>
                    <description>Action to trigger when the barcode has been successfully detected.</description>
                </property>
            </propertyGroup>

            <propertyGroup caption="Common">
                <systemProperty key="Name" />
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Dimensions">
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="percentage">
                    <caption>Width unit</caption>
                    <description>Percentage: portion of parent size. Pixels: absolute amount of pixels.</description>
                    <enumerationValues>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="width" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description />
                </property>
                <property key="heightUnit" type="enumeration" defaultValue="percentageOfWidth">
                    <caption>Height unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="percentageOfWidth">Percentage of width</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentageOfParent">Percentage of parent</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="height" type="integer" defaultValue="75">
                    <caption>Height</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <property key="detectionLogic" type="enumeration" defaultValue="native">
                <caption>Detection logic</caption>
                <description>Choose the detection logic to use for barcode scanning.</description>
                <enumerationValues>
                    <enumerationValue key="zxing">ZXing</enumerationValue>
                    <enumerationValue key="native">BarcodeDetector API (experimental fast scan, fallback to ZXing)</enumerationValue>
                </enumerationValues>
            </property>
        </propertyGroup>
    </properties>
</widget>
