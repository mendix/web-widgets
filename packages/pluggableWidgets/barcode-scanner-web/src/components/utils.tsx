import {
    BinaryBitmap,
    BrowserMultiFormatReader,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    BarcodeFormat,
    DecodeHintType,
    Result
} from "@zxing/library";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export function decodeBarcodeDetector(
    resolve: (value: Result | string) => void,
    reject: (reason?: Error) => void,
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement,
    barcodeDetector?: any
): void {
    const croppedOnCanvas = drawCropOnCanvas(captureCanvas, videoRef, canvasMiddle);
    barcodeDetector
        .detect(croppedOnCanvas)
        .then((barcodes: any) => {
            if (barcodes[0]?.rawValue) {
                resolve(barcodes[0]?.rawValue);
            } else {
                setTimeout(
                    () =>
                        decodeBarcodeDetector(resolve, reject, captureCanvas, videoRef, canvasMiddle, barcodeDetector),
                    50
                );
            }
        })
        .catch((err: any) => {
            reject(err);
        });
}

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
