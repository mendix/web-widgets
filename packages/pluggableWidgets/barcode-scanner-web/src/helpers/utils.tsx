import {
    BarcodeFormat,
    BinaryBitmap,
    BrowserMultiFormatReader,
    DecodeHintType,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    Result
} from "@zxing/library";
import { RefObject } from "react";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const mediaStreamConstraints: MediaStreamConstraints = {
    audio: false,
    video: {
        facingMode: "environment",
        width: { min: 1280, ideal: 4096, max: 4096 },
        height: { min: 720, ideal: 2160, max: 2160 }
    }
};

export type ReaderProps = {
    onSuccess?: (data: string) => void;
    onError?: (e: Error) => void;
    useCrop: boolean;
    barcodeFormats?: BarcodeFormatsType[];
    useAllFormats: boolean;
    canvasMiddleRef: RefObject<HTMLDivElement>;
    detectionLogic?: "zxing" | "native";
};

export type UseReaderHook = (args: ReaderProps) => { ref: RefObject<HTMLVideoElement>; useBrowserAPI: boolean };

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
