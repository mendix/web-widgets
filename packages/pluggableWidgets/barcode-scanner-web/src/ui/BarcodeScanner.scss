.mx-barcode-scanner {
    position: relative;

    .mx-barcode-scanner-content {
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background: #000000;
        // Should be higher than the z-index of `.layout-atlas .region-sidebar` so it sits on top of the page.
        z-index: 110;

        .video {
            position: relative;
            width: 100%;
            height: 100%;
            object-fit: cover;
        }

        .video-canvas {
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 100%;
            display: flex;
            flex-direction: row;
            align-items: stretch;

            .canvas-background {
                background-color: rgba(0, 0, 0, 0.7);
            }

            .canvas-left {
                flex: 0;
                flex-basis: 33.333%;
            }

            $screen-md: 768px;
            $screen-lg: 992px;
            $screen-xl: 1200px;
            .canvas-middle {
                flex: 0;
                flex-basis: 33.333%;
                display: flex;
                flex-direction: column;

                .canvas-middle-top {
                    flex: 1;
                }

                .canvas-middle-middle {
                    flex: 0;
                    padding-top: 100%;
                    position: relative;
                    $corner-offset: 13px;
                    .corner {
                        position: absolute;
                        width: 32px;
                        height: 32px;
                        box-sizing: border-box;
                        $corner-border: 5px solid #ffffff;
                        border-left: $corner-border;
                        border-top: $corner-border;

                        &.corner-top-left {
                            top: -$corner-offset;
                            left: -$corner-offset;
                        }
                        &.corner-top-right {
                            top: -$corner-offset;
                            right: -$corner-offset;
                            transform: rotate(90deg);
                        }
                        &.corner-bottom-right {
                            bottom: -$corner-offset;
                            right: -$corner-offset;
                            transform: rotate(180deg);
                        }
                        &.corner-bottom-left {
                            bottom: -$corner-offset;
                            left: -$corner-offset;
                            transform: rotate(270deg);
                        }
                    }
                }

                .canvas-middle-bottom {
                    flex: 1;
                }
            }

            .canvas-right {
                flex: 0;
                flex-basis: 33.333%;
            }
        }
    }

    &.debug {
        // for debug purposes, we need to determine if barcode scanner is using native or zxing
        &.mx-barcode-detector {
            .mx-barcode-scanner-content {
                border: 4px solid var(--brand-success, #16aa16);
            }
        }
    }
}

// Overwrite `atlas_core/web/core/_legacy/_mxui.scss` for this particular widget because otherwise
// iOS Safari will in certain cases put the top and/or bottom bar on top of the overlay of this widget.
.mx-dataview-content,
.mx-scrollcontainer-wrapper:not(.mx-scrollcontainer-nested) {
    -webkit-overflow-scrolling: auto !important;
}
