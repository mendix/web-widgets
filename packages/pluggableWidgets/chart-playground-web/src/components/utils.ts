function isNonEmptyString(value: undefined | null | string): value is string {
    return value !== null && value !== undefined && value !== "";
}

export const emptyObjectString = "{}";

export function ifNonEmptyStringElseEmptyObjectString(value: undefined | null | string): string {
    return isNonEmptyString(value) ? value : emptyObjectString;
}
