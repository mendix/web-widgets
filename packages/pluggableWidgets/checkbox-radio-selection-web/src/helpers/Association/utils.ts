import {
    ActionValue,
    DynamicValue,
    ListAttributeValue,
    ListExpressionValue,
    ListValue,
    ListWidgetValue,
    ReferenceSetValue,
    ReferenceValue,
    ValueStatus
} from "mendix";
import {
    CheckboxRadioSelectionContainerProps,
    OptionsSourceCustomContentTypeEnum
} from "../../../typings/CheckboxRadioSelectionProps";

export function extractAssociationProps(
    props: CheckboxRadioSelectionContainerProps
): [
    ReferenceValue | ReferenceSetValue,
    ListValue,
    ListAttributeValue<string> | ListExpressionValue<string>,
    DynamicValue<string>,
    boolean,
    ActionValue | undefined,
    ListWidgetValue | undefined,
    OptionsSourceCustomContentTypeEnum
] {
    const attr = props.attributeAssociation;
    const ds = props.optionsSourceAssociationDataSource!;

    // Determine caption provider based on caption type
    const captionProvider =
        props.optionsSourceAssociationCaptionType === "attribute"
            ? props.optionsSourceAssociationCaptionAttribute!
            : props.optionsSourceAssociationCaptionExpression!;

    // For simplicity, we'll create a basic empty option
    const emptyOption: DynamicValue<string> = {
        status: ValueStatus.Available,
        value: ""
    };

    // Checkbox Radio Selection controls don't need clearable like combobox does
    const clearable = false;

    const onChangeEvent = props.onChangeEvent;
    const customContent = props.optionsSourceAssociationCustomContent;
    const customContentType = props.optionsSourceCustomContentType;

    return [attr, ds, captionProvider, emptyOption, clearable, onChangeEvent, customContent, customContentType];
}
