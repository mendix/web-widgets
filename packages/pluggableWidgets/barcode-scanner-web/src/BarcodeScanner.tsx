import { executeAction } from "@mendix/widget-plugin-platform/framework/execute-action";
import { ValueStatus } from "mendix";
import { createElement, FunctionComponent, useCallback } from "react";
import { BarcodeScannerContainerProps } from "../typings/BarcodeScannerProps";
import { BarcodeScanner as BarcodeScannerComponent } from "./components/BarcodeScanner";

export const BarcodeScanner: FunctionComponent<BarcodeScannerContainerProps> = props => {
    const onDetect = useCallback(
        (data: string) => {
            if (props.datasource?.status !== ValueStatus.Available) {
                return;
            }
            if (data !== props.datasource.value) {
                props.datasource.setValue(data);
            }
            executeAction(props.onDetect);
        },
        [props.onDetect, props.datasource]
    );
    return (
        <BarcodeScannerComponent
            onDetect={onDetect}
            showMask={props.showMask}
            class={props.class}
            useAllFormats={props.useAllFormats}
            barcodeFormats={props.barcodeFormats}
            heightUnit={props.heightUnit}
            height={props.height}
            widthUnit={props.widthUnit}
            width={props.width}
            detectionLogic={props.detectionLogic}
        />
    );
};
