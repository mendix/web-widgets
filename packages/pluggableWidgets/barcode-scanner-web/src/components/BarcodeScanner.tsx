import { createElement, ReactElement, ReactNode, useCallback, SyntheticEvent, useRef, RefObject } from "react";
import classNames from "classnames";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { Dimensions, getDimensions } from "@mendix/widget-plugin-platform/utils/get-dimensions";
import { useCustomErrorMessage } from "../hooks/useCustomErrorMessage";
import { useReader } from "../hooks/useReader";
import { BarcodeFormatsType } from "../../typings/BarcodeScannerProps";

import "../ui/BarcodeScanner.scss";

interface BarcodeScannerOverlayProps extends Dimensions {
    showMask: boolean;
    class: string;
    children?: ReactNode;
    canvasMiddleMiddleRef?: RefObject<HTMLDivElement>;
}

export function BarcodeScannerOverlay({
    children,
    class: className,
    showMask,
    canvasMiddleMiddleRef,
    ...dimensions
}: BarcodeScannerOverlayProps): ReactElement {
    return (
        <div className={classNames("mx-barcode-scanner", className)} style={getDimensions(dimensions)}>
            <div className={classNames("mx-barcode-scanner-content")}>
                {children}
                {showMask ? (
                    <div className={classNames("video-canvas")}>
                        <div className={classNames("canvas-left", "canvas-background")} />
                        <div className={classNames("canvas-middle")}>
                            <div className={classNames("canvas-middle-top", "canvas-background")} />
                            <div
                                ref={canvasMiddleMiddleRef}
                                id={"canvas-middle-middle"}
                                className={classNames("canvas-middle-middle")}
                            >
                                <div className={classNames("corner", "corner-top-left")} />
                                <div className={classNames("corner", "corner-top-right")} />
                                <div className={classNames("corner", "corner-bottom-right")} />
                                <div className={classNames("corner", "corner-bottom-left")} />
                            </div>
                            <div className={classNames("canvas-middle-bottom", "canvas-background")} />
                        </div>
                        <div className={classNames("canvas-right", "canvas-background")} />
                    </div>
                ) : null}
            </div>
        </div>
    );
}

export interface BarcodeScannerProps extends Dimensions {
    onDetect?: (data: string) => void;
    showMask: boolean;
    class: string;
    useAllFormats: boolean;
    barcodeFormats?: BarcodeFormatsType[];
}

export function BarcodeScanner({
    onDetect,
    showMask,
    class: className,
    barcodeFormats,
    useAllFormats,
    ...dimensions
}: BarcodeScannerProps): ReactElement | null {
    const [errorMessage, setError] = useCustomErrorMessage();
    const canvasMiddleRef = useRef<HTMLDivElement>(null);
    const videoRef = useReader({
        onSuccess: onDetect,
        onError: setError,
        useCrop: showMask,
        barcodeFormats,
        useAllFormats,
        canvasMiddleRef
    });
    const supportsCameraAccess = typeof navigator?.mediaDevices?.getUserMedia === "function";
    const onCanPlay = useCallback((event: SyntheticEvent<HTMLVideoElement>) => {
        if (event.currentTarget.paused) {
            event.currentTarget.play();
        }
    }, []);

    if (!supportsCameraAccess) {
        // Mendix ensures that Mendix apps are only run in the supported browsers and all of them
        // support the `navigator.mediaDevices.getUserMedia` API. So no additional error handling
        // needs to be done, but just in case we soft catch it.
        return (
            <Alert bootstrapStyle="danger">
                The barcode scanner widget is only compatible with certain browsers and requires a secure HTTPS
                connection in certain browsers. If you encounter this error message as an user, please contact your
                system administrator. If you are a Mendix developer, please refer to the appropriate docs on how to
                resolve this issue.
            </Alert>
        );
    }

    if (errorMessage) {
        return <Alert bootstrapStyle="danger">{errorMessage}</Alert>;
    }

    return (
        <BarcodeScannerOverlay
            class={className}
            showMask={showMask}
            canvasMiddleMiddleRef={canvasMiddleRef}
            {...dimensions}
        >
            <video className="video" ref={videoRef} onCanPlay={onCanPlay} />
        </BarcodeScannerOverlay>
    );
}
