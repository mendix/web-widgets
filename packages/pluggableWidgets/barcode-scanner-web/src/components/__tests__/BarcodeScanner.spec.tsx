import "@testing-library/jest-dom";
import { act, render, screen, waitFor } from "@testing-library/react";
import { Dimensions } from "@mendix/widget-plugin-platform/utils/get-dimensions";
import { NotFoundException } from "@zxing/library/cjs";
import { BarcodeScanner } from "../BarcodeScanner";

let useReaderMock = jest.fn();
jest.mock("../../hooks/useReader", () => ({
    // we can't use mock directly because of variable hosting.
    useReader: (...args: any[]) => useReaderMock(...args)
}));

describe("Barcode scanner", () => {
    const backupMediaDevices = window.navigator.mediaDevices;
    const dimensions: Dimensions = {
        widthUnit: "percentage",
        width: 100,
        heightUnit: "percentageOfParent",
        height: 100
    };

    function mockGetUserMedia(getUserMediaMock: jest.Mock): void {
        Object.defineProperty(window.navigator, "mediaDevices", {
            value: { getUserMedia: getUserMediaMock },
            writable: true
        });
    }

    afterEach(() => {
        // reset the mocking
        Object.defineProperty(window.navigator, "mediaDevices", {
            value: backupMediaDevices,
            writable: true
        });
        useReaderMock = jest.fn();
    });

    it("renders video and overlay correctly", () => {
        mockGetUserMedia(jest.fn());
        expect(
            render(<BarcodeScanner useAllFormats class="" showMask {...dimensions} />).asFragment()
        ).toMatchSnapshot();
    });

    it("does not show the overlay when the user opts out of it", () => {
        mockGetUserMedia(jest.fn());
        expect(
            render(<BarcodeScanner useAllFormats class="" showMask={false} {...dimensions} />).asFragment()
        ).toMatchSnapshot();
    });

    it("shows an appropriate error when the mediaDevices API is not present (like over http)", async () => {
        expect(navigator.mediaDevices).toBe(undefined);
        expect(
            render(<BarcodeScanner useAllFormats class="" showMask {...dimensions} />).asFragment()
        ).toMatchSnapshot();
    });

    it("prop health check: pass onDetect prop as onSuccess callback", async () => {
        const onDetectMock = jest.fn();
        useReaderMock.mockImplementationOnce((args: any) => {
            setTimeout(() => args.onSuccess("42"), 100);
        });
        mockGetUserMedia(jest.fn());

        render(<BarcodeScanner useAllFormats class="" onDetect={onDetectMock} showMask {...dimensions} />);

        await waitFor(() => expect(onDetectMock).toHaveBeenCalledWith("42"));
    });

    describe("shows an appropriate error to the user", () => {
        it("in the form of text when a generic error occurs", async () => {
            useReaderMock.mockImplementationOnce((args: any) => {
                setTimeout(() => args.onError(new Error("some error message")), 100);
            });
            mockGetUserMedia(jest.fn());

            await act(async () => {
                render(<BarcodeScanner useAllFormats class="" showMask {...dimensions} />);
            });
            await waitFor(() => expect(screen.getByText(/some error message/i)).toBeVisible());
        });

        it("in the form of text when the code scanner unexpectedly fails", async () => {
            useReaderMock.mockImplementationOnce((args: any) => {
                setTimeout(() => {
                    args.onError(new NotFoundException("Unable to decode from stream"));
                }, 100);
            });

            mockGetUserMedia(jest.fn());

            await act(async () => {
                render(<BarcodeScanner useAllFormats class="" showMask {...dimensions} />);
            });
            await waitFor(() => expect(screen.getByText(/Unable to decode from stream/i)).toBeVisible());
        });
    });
});
