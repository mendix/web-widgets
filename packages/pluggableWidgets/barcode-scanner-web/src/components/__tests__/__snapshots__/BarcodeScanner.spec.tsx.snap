// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Barcode scanner does not show the overlay when the user opts out of it 1`] = `
<DocumentFragment>
  <div
    class="mx-barcode-scanner mx-zxing-detector"
    style="width: 100%; height: 100%;"
  >
    <div
      class="mx-barcode-scanner-content"
    >
      <video
        class="video"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Barcode scanner renders video and overlay correctly 1`] = `
<DocumentFragment>
  <div
    class="mx-barcode-scanner mx-zxing-detector"
    style="width: 100%; height: 100%;"
  >
    <div
      class="mx-barcode-scanner-content"
    >
      <video
        class="video"
      />
      <div
        class="video-canvas"
      >
        <div
          class="canvas-left canvas-background"
        />
        <div
          class="canvas-middle"
        >
          <div
            class="canvas-middle-top canvas-background"
          />
          <div
            class="canvas-middle-middle"
            id="canvas-middle-middle"
          >
            <div
              class="corner corner-top-left"
            />
            <div
              class="corner corner-top-right"
            />
            <div
              class="corner corner-bottom-right"
            />
            <div
              class="corner corner-bottom-left"
            />
          </div>
          <div
            class="canvas-middle-bottom canvas-background"
          />
        </div>
        <div
          class="canvas-right canvas-background"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Barcode scanner shows an appropriate error when the mediaDevices API is not present (like over http) 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
    The barcode scanner widget is only compatible with certain browsers and requires a secure HTTPS connection in certain browsers. If you encounter this error message as an user, please contact your system administrator. If you are a Mendix developer, please refer to the appropriate docs on how to resolve this issue.
  </div>
</DocumentFragment>
`;
